faucet/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + encointer_reputation_commitments::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type Currency: Currency<Self::AccountId>;
+		type FaucetMaster: EnsureOrigin<Self::RuntimeOrigin>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Transfer some balance to another account.
+		#[pallet::call_index(0)]
+		#[pallet::weight(10_000)]
+		pub fn drip(
+			origin: OriginFor<T>,
+			cid: CommunityIdentifier,
+			cindex: CeremonyIndexType,
+		) -> DispatchResultWithPostInfo {
+			let from = ensure_signed(origin)?;
+
+			<encointer_reputation_commitments::Pallet<T>>::do_commit_reputation(
+				&from,
+				cid,
+				cindex,
+				Self::reputation_commitments_purpose_id(),
+				None,
+			)?;
+
+			T::Currency::transfer(&Self::account_id(), &from, Self::drip_amount(), KeepAlive)
+				.map_err(|_| <Error<T>>::FaucetEmpty)?;
+
+			Self::deposit_event(Event::Dripped(from, Self::drip_amount()));
+
+			Ok(().into())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(10_000)]
+		pub fn set_drip_amount(
+			origin: OriginFor<T>,
+			drip_amount: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			T::FaucetMaster::ensure_origin(origin)?;
+			<DripAmount<T>>::put(drip_amount);
+			info!(target: LOG, "set drip amount to {:?}", drip_amount);
+			Self::deposit_event(Event::DripAmountUpdated(drip_amount));
+			Ok(().into())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// The account ID of the faucet.
+		///
+		/// This actually does computation. If you need to keep using it, then make sure you cache the
+		/// value and only call this once.
+		pub fn account_id() -> T::AccountId {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		pub fn pot() -> BalanceOf<T> {
+			T::Currency::free_balance(&Self::account_id())
+				// Must never be less than 0 but better be safe.
+				.saturating_sub(T::Currency::minimum_balance())
+		}
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub drip_amount: BalanceOf<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { drip_amount: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			// Create faucet account
+			let account_id = <Pallet<T>>::account_id();
+			let min = T::Currency::minimum_balance();
+			if T::Currency::free_balance(&account_id) < min {
+				let _ = T::Currency::make_free_balance_be(&account_id, min);
+			}
+			<DripAmount<T>>::put(self.drip_amount);
+
+			<encointer_reputation_commitments::Pallet<T>>::do_register_purpose(
+				DescriptorType::from_str("EncointerFaucet").unwrap(),
+			)
+			.expect("In case of purpose registry overflow, we cannot use this pallet.");
```

Comment:
the problem with setting this up at genesis is that this will fail for runtime upgrades from earlier versions. So either we implement state migrations that initially link a purpose and a faucet or we find another solution.
As encointer has no root origin, we have no way of setting this link manually

Reply:
an alternative design approach would be:
the key for the StorageMap at `reputation_commitments::Commitments` can be either an integer like now or it is the account_id of the faucet and therefore derived automatically and uniquely and exclusively tied to a specific pot

```
enum PurposeIdType {
    external(u32),
    internal(AccountId),
}
```
this way, no registration is needed for internal/onchain use. you'd obviously need to adjust your check of a purpose exists because that ensure is only needed for external purposes

Reply:
i dont fully understand the issue with the genesis config. can you explain?

the second approach seems too specific to me, because we also need internal uses that have an u32 id, like for example in democracy.